src/Bonus.java

```diff
@@ -1,66 +1,59 @@
+        Random random = new Random();
+        HashSet<Integer> set = new HashSet<>();
+        while (set.size() < 4) {
+            set.add(random.nextInt(10));
+        }
+        return set;
     }
 
-    public static void/*moet dit returntype "void" zijn of wat anders?*/ setToStringConverter(/*Heeft deze methode nog parameter(s) nodig?*/){
-        /*
-        Vul hier de body van de methode in.
-
-        Stappenplan:
-        - Maak als eerst een String variabele met als waarde een lege String. (of gebruik een StringBuilder)
-        - Schrijf vervolgens een for-loop om de items in de hashset een voor een aan de String variabele toe te voegen.
-        - Return de (gevulde) String variabele
-         */
+    public static String setToStringConverter(HashSet<Integer> secretNumber) {
+        String string = "";
+        for (int i = 0; i < secretNumber.size(); i++) {
+            string += secretNumber.toArray()[i];
+        }
+        return string;
     }
 
-
-
-    public static void/*moet dit "void" zijn of wat anders?*/ feedback(String stringnumber) {
+    public static void feedback(String stringNumber) {
         Scanner scanner = new Scanner(System.in);
-        StringBuilder feedback = new StringBuilder();
-        System.out.println("+ = juiste nummer op de juiste plek, O = juiste nummer verkeerde plek, X = verkeerde nummer");
-        System.out.println("Doe een gok, Let op vul 4 getallen in.");
+        System.out.println("Welkom bij Mastermind!");
+        System.out.println("Doe een gok. Let op: vul 4 getallen in.");
+        System.out.println("Uitleg van de feedback: \n+ = juiste nummer op de juiste plek, O = juiste nummer verkeerde plek, X = verkeerde nummer");
         String guess = scanner.nextLine();
-        if (Objects.equals(guess, stringnumber)) {
-            System.out.println("gefeliciteerd je hebt het goed");
-        } else {
+        while (!Objects.equals(guess, stringNumber)) {
+            StringBuilder feedback = new StringBuilder();
             for (int i = 0; i < 4; i++) {
-                if (guess.substring(i, i + 1).equals(stringnumber.substring(i, i + 1))) {
+                if (guess.substring(i, i + 1).equals(stringNumber.substring(i, i + 1))) {
                     feedback.append("+");
-                } else if (stringnumber.contains(guess.substring(i, i + 1))) {
+                } else if (stringNumber.contains(guess.substring(i, i + 1))) {
                     feedback.append("0");
                 } else {
                     feedback.append("X");
                 }
             }
+            System.out.println(feedback.toString());
+            System.out.println("Try again...");
+            guess = scanner.nextLine();
         }
-        System.out.println(feedback.toString());
+        System.out.println("Je hebt het correct!");
     }
 }
+
+// Antwoord stap 4: Omdat we een vast aantal cijfers nodig hadden, dat gaat ook met een Array.
```

Comment:
Ik tast hier ook nog in het duister hoor met bv Vector?? als alternatief, maar ik vermoedde al wel dat ArrayList een logische optie zou zijn. Copilot vind hier ook het "hette" van:

Hier is een meer gedetailleerde uitleg over waarom een `ArrayList` de betere keuze zou zijn, en waarom een `Array` niet ideaal is:

### Waarom `ArrayList` de beste optie is:

1. **Flexibiliteit in Grootte**:
   - **ArrayList**: Kan dynamisch groeien en krimpen, wat betekent dat je gemakkelijk elementen kunt toevoegen zonder je zorgen te maken over de vaste grootte.
   - **Array**: Heeft een vaste grootte, wat betekent dat je de grootte moet bepalen bij het creëren. Dit maakt het minder flexibel als je later meer of minder elementen wilt toevoegen.

2. **Uniciteit**:
   - **HashSet**: Zorgt ervoor dat alle elementen uniek zijn. Voor de Mastermind-opdracht is het niet noodzakelijk dat de nummers uniek zijn, dus dit is overbodige functionaliteit.
   - **ArrayList**: Staat duplicaten toe, wat acceptabel is in de context van de Mastermind-opdracht.

3. **Toegang tot Elementen**:
   - **ArrayList**: Biedt eenvoudige methoden voor het toevoegen, verwijderen en ophalen van elementen op basis van indexen, wat de manipulatie van de lijst vergemakkelijkt.
   - **Array**: Biedt directe toegang op basis van indexen, maar het toevoegen en verwijderen van elementen is minder flexibel en vereist meer handmatige inspanning.

4. **Geheugenbeheer**:
   - **HashSet**: Heeft een hogere geheugenoverhead vanwege de interne hashingmechanismen die worden gebruikt om uniciteit te garanderen.
   - **ArrayList**: Heeft een lagere geheugenoverhead en is eenvoudiger omdat het gewoon een dynamische array is die elementen in volgorde opslaat.

### Waarom `Array` niet de juiste oplossing is:

- **Vaste Grootte**: Een `Array` heeft een vaste grootte, wat betekent dat je van tevoren moet weten hoeveel elementen je nodig hebt. Dit maakt het moeilijker om flexibel te zijn met het aantal elementen.
- **Minder Flexibel**: Het toevoegen en verwijderen van elementen in een array is minder flexibel en vereist meer handmatige inspanning in vergelijking met een `ArrayList`.

### Alternatieve List Types:

1. **LinkedList**:
   - **Voordelen**: Goed voor scenario's waar veel toevoegen en verwijderen van elementen in het begin of einde van de lijst nodig is.
   - **Nadelen**: Heeft een hogere overhead voor toegang tot elementen op basis van indexen in vergelijking met `ArrayList`.

2. **Vector**:
   - **Voordelen**: Synchronized, wat betekent dat het thread-safe is.
   - **Nadelen**: Heeft een performance overhead door synchronisatie en is meestal niet nodig voor niet-concurrente scenario's.

### Verbeterde Implementatie met `ArrayList`:

```java
public static List<Integer> randomNumberGenerator() {
    Random random = new Random();
    List<Integer> list = new ArrayList<>();
    while (list.size() < 4) {
        list.add(random.nextInt(10)); // Genereren van getallen van 0 tot 9
    }
    return list;
}
```

### Commentaar op het gebruik van `HashSet`:

```java
// Antwoord stap 4: HashSet is geen goede keuze omdat het uniekheid afdwingt, wat hier niet nodig is.
// Bovendien zorgt HashSet voor extra geheugenoverhead door hashingmechanismen.
// Een ArrayList is een betere keuze vanwege zijn flexibiliteit in grootte en eenvoudiger geheugenbeheer.
```

Door deze aanpassingen en uitleg wordt duidelijk waarom een `ArrayList` de voorkeur heeft boven een `HashSet` en een `Array` voor deze specifieke opdracht.

Reply:
En wat betreft waarom dan Arraylist ipv HashSet ook nog wat uitgebreidere uitleg gekregen:

Hier is een meer gedetailleerde uitleg over waarom een `ArrayList` een betere keuze zou zijn dan een `HashSet` in dit geval:

1. **Uniciteit vs. Duplikaat Elementen**:
   - **HashSet**: Zorgt ervoor dat alle elementen uniek zijn. Dit is niet altijd gewenst, vooral niet als duplicaten zijn toegestaan in de context van de applicatie.
   - **ArrayList**: Staat duplicaten toe en is meer flexibel in situaties waar dezelfde waarde meerdere keren kan voorkomen.

2. **Complexiteit en Geheugen**:
   - **HashSet**: Heeft een hogere geheugenoverhead vanwege de interne hashingmechanismen die worden gebruikt om uniciteit te garanderen. Dit kan onnodige complexiteit en geheugenverbruik toevoegen, vooral bij kleinere datasets.
   - **ArrayList**: Heeft een lagere geheugenoverhead en is eenvoudiger, omdat het gewoon een dynamische array is die elementen in volgorde opslaat.

3. **Prestaties**:
   - **HashSet**: Heeft gemiddeld O(1) tijdcomplexiteit voor invoegen en zoeken, maar dit voordeel komt met de kosten van hogere geheugenverbruik en complexiteit.
   - **ArrayList**: Heeft O(1) tijdcomplexiteit voor het toevoegen van elementen (amortized) en O(n) voor zoeken, maar dit is vaak voldoende voor kleinere datasets en biedt een eenvoudigere implementatie.

4. **Dynamische Groei**:
   - **ArrayList**: Kan eenvoudig worden uitgebreid wanneer meer elementen moeten worden toegevoegd. Het past automatisch de grootte van de interne array aan wanneer dat nodig is.
   - **HashSet**: Groeit ook dynamisch, maar met meer overhead door herhashing en het handhaven van de unieke-eigenschap.

5. **Gebruiksgemak**:
   - **ArrayList**: Biedt een eenvoudigere API voor toegang tot elementen via indexen, wat handig kan zijn voor iteraties en specifieke toegangspatronen.
   - **HashSet**: Heeft geen index-gebaseerde toegang, wat het minder flexibel maakt voor bepaalde operaties die afhankelijk zijn van volgorde of positie.

In de context van de Mastermind-applicatie, waar het belangrijk is om een reeks van getallen te genereren en te manipuleren zonder de noodzaak van uniekheid, biedt een `ArrayList` een eenvoudiger en efficiënter alternatief dan een `HashSet`.